internal/ls/completions.go

```diff
@@ -4543,3 +4552,289 @@ func getArgumentInfoForCompletions(node *ast.Node, position int, file *ast.Sourc
 		argumentCount: info.argumentCount,
 	}
 }
+
+type itemData struct {
+	FileName   string          `json:"fileName"`
+	Position   int             `json:"position"`
+	Source     string          `json:"source,omitempty"`
+	Name       string          `json:"name,omitempty"`
+	AutoImport *autoImportData `json:"autoImport,omitempty"`
+}
+
+// !!! CompletionEntryDataAutoImport
+type autoImportData struct{}
+
+// Special values for `CompletionInfo['source']` used to disambiguate
+// completion items with the same `name`. (Each completion item must
+// have a unique name/source combination, because those two fields
+// comprise `CompletionEntryIdentifier` in `getCompletionEntryDetails`.
+//
+// When the completion item is an auto-import suggestion, the source
+// is the module specifier of the suggestion. To avoid collisions,
+// the values here should not be a module specifier we would ever
+// generate for an auto-import.
+const (
+	// Completions that require `this.` insertion text
+	SourceThisProperty = "ThisProperty/"
+	// Auto-import that comes attached to a class member snippet
+	SourceClassMemberSnippet = "ClassMemberSnippet/"
+	// A type-only import that needs to be promoted in order to be used at the completion location
+	SourceTypeOnlyAlias = "TypeOnlyAlias/"
+	// Auto-import that comes attached to an object literal method snippet
+	SourceObjectLiteralMethodSnippet = "ObjectLiteralMethodSnippet/"
+	// Case completions for switch statements
+	SourceSwitchCases = "SwitchCases/"
+	// Completions for an object literal expression
+	SourceObjectLiteralMemberWithComma = "ObjectLiteralMemberWithComma/"
+)
+
+func (l *LanguageService) ResolveCompletionItem(
+	ctx context.Context,
+	item *lsproto.CompletionItem,
+	data *itemData,
+	clientOptions *lsproto.CompletionClientCapabilities,
+	preferences *UserPreferences,
+) (*lsproto.CompletionItem, error) {
+	if data == nil {
+		return nil, errors.New("completion item data is nil")
+	}
+
+	program, file := l.tryGetProgramAndFile(data.FileName)
+	if file == nil {
+		return nil, fmt.Errorf("file not found: %s", data.FileName)
+	}
+
+	return l.getCompletionItemDetails(ctx, program, data.Position, file, item, data, clientOptions, preferences), nil
+}
+
+func GetCompletionItemData(item *lsproto.CompletionItem) (*itemData, error) {
+	bytes, err := json.Marshal(item.Data)
```

Comment:
To avoid this, we should probably special case some of these incoming `any`s as `json.RawMessage`, maybe?

Reply:
Yes, that would be nice. I can try and add this to the lsp generator in a separate PR.

Reply:
The annoying bit is that it would probably break things where we are passing the objects as-is, I think

Reply:
Right, because then the type of `CompletionItem.Data` would be bytes and we'd have to marshall that as we construct the object.